File: GESDISC/OMI-Aura_L2G-OMSO2G_2017m0123_v003-2017m0124t071057.he5.py
"""

This example code illustrates how to access and visualize a GESDISC OMI L2G 
HDF-EOS5 Grid file in Python.

If you have any questions, suggestions, or comments on this example, please use
the HDF-EOS Forum (http://hdfeos.org/forums).  If you would like to see an
example of any other NASA HDF/HDF-EOS data product that is not listed in the
HDF-EOS Comprehensive Examples page (http://hdfeos.org/zoo), feel free to
contact us at eoshelp@hdfgroup.org or post it at the HDF-EOS Forum
(http://hdfeos.org/forums).

Usage:  save this script and run

  $python OMI-Aura_L2G-OMSO2G_2017m0123_v003-2017m0124t071057.he5.py

The HDF-EOS5 file must be in your current working directory.

Tested under: Python 2.7.13 :: Anaconda 4.3.22 
Last updated: 2017-12-19
"""
import os

import matplotlib as mpl
import matplotlib.pyplot as plt
from mpl_toolkits.basemap import Basemap
import numpy as np
import h5py

FILE_NAME = 'OMI-Aura_L2G-OMSO2G_2017m0123_v003-2017m0124t071057.he5'    


with h5py.File(FILE_NAME, mode='r') as f:
    PATH_NAME = '/HDFEOS/GRIDS/OMI Total Column Amount SO2/Data Fields/'
    DATAFIELD_NAME = PATH_NAME + 'ColumnAmountSO2_PBL'
    LAT_NAME = PATH_NAME + 'Latitude'
    LON_NAME = PATH_NAME + 'Longitude'

    dset = f[DATAFIELD_NAME]
    data = dset[:]

    # Read lat/lon data.
    lat = f[LAT_NAME][:]
    lon = f[LON_NAME][:]
    
    # Get attributes needed for the plot.
    # String attributes actually come in as the bytes type and should
    # be decoded to UTF-8 (python3).
    title = dset.attrs['Title'].decode()
    units = dset.attrs['Units'].decode()
    valid_range = dset.attrs['ValidRange']
    _FillValue = dset.attrs['_FillValue']
    add_offset = dset.attrs['Offset']
    scale_factor = dset.attrs['ScaleFactor']

    # Handle fill value.
    data[data == _FillValue] = np.nan
    data = data * scale_factor + add_offset

    # Handle valid range.
    data[data < valid_range[0]] = np.nan
    data[data > valid_range[1]] = np.nan

    data = np.ma.masked_where(np.isnan(data), data)
    # Subset data at nCandidate = 0
    data = data[0,:,:]
    lon = lon[0,:,:]
    lat = lat[0,:,:]

    # Draw an equidistant cylindrical projection using the low resolution
    # coastline database.
    m = Basemap(projection='cyl', resolution='l',
                llcrnrlat=-90, urcrnrlat = 90,
                llcrnrlon=-180, urcrnrlon = 180)
    m.drawcoastlines(linewidth=0.5)
    m.drawparallels(np.arange(-90., 120., 30.), labels=[1, 0, 0, 0])
    m.drawmeridians(np.arange(-180, 180., 45.), labels=[0, 0, 0, 1])
    m.scatter(lon, lat, c=data, s=0.1, cmap=plt.cm.jet,
              edgecolors=None, linewidth=0)
    cb = m.colorbar()
    cb.set_label(units)

    basename = os.path.basename(FILE_NAME)
    plt.title('{0}\n{1}'.format(basename, title+' at nCandidate=0'))

    fig = plt.gcf()
    pngfile = "{0}.py.png".format(basename)
    fig.savefig(pngfile)

File: GESDISC/OMI_Aura_L2G_OMSO2G_2017m0123_v003_2017m0124t071057_he5.m
%
%  This example code illustrates how to access and visualize OMI Grid file
% in MATLAB. 
%
%  If you have any questions, suggestions, comments on this example, please 
% use the HDF-EOS Forum (http://hdfeos.org/forums). 
% 
%  If you would like to see an  example of any other NASA HDF/HDF-EOS data 
% product that is not listed in the HDF-EOS Comprehensive Examples page 
% (http://hdfeos.org/zoo), feel free to contact us at eoshelp@hdfgroup.org 
% or post it at the HDF-EOS Forum (http://hdfeos.org/forums).
%
% Usage:save this script and run (without .m at the end)
%                                   
%
% $matlab -nosplash -nodesktop -r OMI_Aura_L2G_OMSO2G_2017m0123_v003_2017m0124t071057_he5
%
% Tested under: MATLAB R2017a
% Last updated: 2017-12-19

clear
% Open the HDF5 File.
FILE_NAME = 'OMI-Aura_L2G-OMSO2G_2017m0123_v003-2017m0124t071057.he5';
file_id = H5F.open (FILE_NAME, 'H5F_ACC_RDONLY', 'H5P_DEFAULT');

% Open the dataset.
DATAFIELD_NAME = ...
'/HDFEOS/GRIDS/OMI Total Column Amount SO2/Data Fields/ColumnAmountSO2_PBL';
data_id = H5D.open (file_id, DATAFIELD_NAME);

Lat_NAME='/HDFEOS/GRIDS/OMI Total Column Amount SO2/Data Fields/Latitude';
lat_id=H5D.open(file_id, Lat_NAME);

Lon_NAME='/HDFEOS/GRIDS/OMI Total Column Amount SO2/Data Fields/Longitude';
lon_id=H5D.open(file_id, Lon_NAME);

% Read the dataset.
data=H5D.read (data_id,'H5T_NATIVE_DOUBLE', 'H5S_ALL', 'H5S_ALL', ...
    'H5P_DEFAULT');

lat=H5D.read(lat_id,'H5T_NATIVE_DOUBLE', 'H5S_ALL', 'H5S_ALL', ...
    'H5P_DEFAULT');

lon=H5D.read(lon_id,'H5T_NATIVE_DOUBLE', 'H5S_ALL', 'H5S_ALL',...
    'H5P_DEFAULT');

% Read the Title.
ATTRIBUTE = 'Title';
attr_id = H5A.open_name (data_id, ATTRIBUTE);
long_name = H5A.read(attr_id, 'H5ML_DEFAULT');

% Read the units.
ATTRIBUTE = 'Units';
attr_id = H5A.open_name (data_id, ATTRIBUTE);
units = H5A.read(attr_id, 'H5ML_DEFAULT');

% Read the offset.
ATTRIBUTE = 'Offset';
attr_id = H5A.open_name (data_id, ATTRIBUTE);
offset = H5A.read(attr_id, 'H5ML_DEFAULT');

% Read the scale.
ATTRIBUTE = 'ScaleFactor';
attr_id = H5A.open_name (data_id, ATTRIBUTE);
scale = H5A.read(attr_id, 'H5ML_DEFAULT');

% Read the fillvalue.
ATTRIBUTE = '_FillValue';
attr_id = H5A.open_name (data_id, ATTRIBUTE);
fillvalue=H5A.read (attr_id, 'H5T_NATIVE_DOUBLE');

% Read the ValidRange.
ATTRIBUTE = 'ValidRange';
attr_id = H5A.open_name (data_id, ATTRIBUTE);
valid_range = H5A.read (attr_id, 'H5T_NATIVE_DOUBLE');

% Read the missingvalue.
ATTRIBUTE = 'MissingValue';
attr_id = H5A.open_name (data_id, ATTRIBUTE);
missingvalue=H5A.read (attr_id, 'H5T_NATIVE_DOUBLE');

attr_id = H5A.open_name (lat_id, ATTRIBUTE);
missingvalue_lat=H5A.read (attr_id, 'H5T_NATIVE_DOUBLE');

attr_id = H5A.open_name (lon_id, ATTRIBUTE);
missingvalue_lon=H5A.read (attr_id, 'H5T_NATIVE_DOUBLE');


% Close and release resources.
H5A.close (attr_id)
H5D.close (lat_id);
H5D.close (lon_id);
H5D.close (data_id);
H5F.close (file_id);

% Replace the fill value with NaN
data(data==fillvalue) = NaN;
lat(lat==missingvalue) = NaN;
lon(lon==missingvalue) = NaN;

% Handle valid range.
data(data < valid_range(1)) = NaN;
data(data > valid_range(2)) = NaN;

% Limit data to 1.0 to 5.0 if you want to match sample image on
% GES DISC website  [1].
% data(data < 0.0) = 0.0;
% data(data > 5.0) = 5.0;

% Apply scale and offset, the equation is scale *(data-offset).
data = scale*(data-offset);

% Subset array.
nCandidate = 1;
data = data(:,:,nCandidate);
lat = lat(:,:,nCandidate);
lon = lon(:,:,nCandidate);

f = figure('Name', FILE_NAME, 'visible', 'off');

axesm('MapProjection','eqdcylin', 'Grid', 'on', 'MeridianLabel', ...
      'on','ParallelLabel','on', 'MLabelParallel','south', ... 
      'FontSize', 7);
coast = load('coast.mat');
cm = colormap('Jet');

% Surfacem is not good for data with many fill values.
% surfacem(lat, lon, data);

% Use scatterm instead.
lat = lat(:)';
lon = lon(:)';
data = data(:)';
scatterm(lat, lon, 1, data);
h = colorbar();

unit = sprintf('%s', units);
set(get(h, 'title'), 'string', unit, 'FontSize', 7);

plotm(coast.lat,coast.long,'k');
tightmap;

DISP_NAME = sprintf('%s', long_name);
title({FILE_NAME; DISP_NAME; ' at nCandidate=0'}, ... 
      'Interpreter', 'None', 'FontSize', 7);
saveas(f, [FILE_NAME '.m.png']);
exit;

% References
% [1] https://disc.gsfc.nasa.gov/datasets/OMSO2G_V003/summary
